### PR TITLE
Handle renaming of local variables in incremental analysis (AST)

### PR DESCRIPTION
**Pull request by TimOrtel**
**This pull requests implements two things for incremental analysis using AST:**

1. Detection of renamed local variables in functions in incremental analysis.
2. Unified output of varinfos. Varinfo names that have been changed are output with their changed names rather than their original names.

***How it works:***
**Detection of renamed local variables:** In compareAST.ml an additional parameter rename_mapping has been added to all functions comparing (parts of) AST nodes. This rename mapping holds assumptions about renamed local variables and is carried through all calls that compare the ASTs. If the assumptions of rename_mapping are never broken and apart from names being modified nothing has changed, the function is guaranteed to not being changed.
The assumptions of rename_mapping are not broken, if all occurrences of variables in the old AST match the occurrences in the new AST, and all variables that have been renamed have the new name in all occurrences in the new AST.

The assumptions of rename_mapping are constructed in compareCil.ml/eqF by looking at the locals before any comparing is performed.

Why is rename_mapping a function attribute and not a global var?
It may be beneficial when rename detection of globals (functions, global vars) is added.

**Unified output of varinfos:**
The name of the old varinfo is now replaced by the name of the new varinfo when a local variable was renamed during an incremental run. 

Additional note: Currently, in renameMapping.ml dn_obj is copied from Cil.dn_obj as Cil does not export dn_obj. This has to be changed in a pull request for Cil.

CFG based comparisons are currently not supported, meaning that the behavior for CFG has not changed.

Please comment on this pull request if you need more information and if you have feedback for me. I will try to work it in as soon as possible.

### CODE REVIEW

File: src/incremental/compareCIL.ml [hunk 1]
```diff
@@ -100,10 +156,22 @@ let compareCilFiles ?(eq=eq_glob) (oldAST: file) (newAST: file) =
...
   let newMap = Cil.foldGlobals newAST addGlobal GlobalMap.empty in
+
+  let global_rename_mapping: method_rename_assumptions = Cil.foldGlobals newAST (fun (current_global_rename_mapping: method_rename_assumption list) global ->
+    match generate_global_rename_mapping oldMap global with
+      | Some rename_mapping -> current_global_rename_mapping @ [rename_mapping]
+      | None -> current_global_rename_mapping
+  ) [] |>
+    (fun mappings -> 
+      let table = Hashtbl.create (List.length mappings) in
+      List.iter (fun mapping -> Hashtbl.add table mapping.original_method_name mapping) mappings;
```
Comment (sim642) [inline src/incremental/compareCIL.ml:167]: Wouldn't `Hashtbl.replace` be enough?

File: src/incremental/compareCIL.ml [hunk 2]
```diff
@@ -100,10 +156,22 @@ let compareCilFiles ?(eq=eq_glob) (oldAST: file) (newAST: file) =
   (* Store a map from functionNames in the old file to the function definition*)
   let oldMap = Cil.foldGlobals oldAST addGlobal GlobalMap.empty in
   let newMap = Cil.foldGlobals newAST addGlobal GlobalMap.empty in
+
+  let global_rename_mapping: method_rename_assumptions = Cil.foldGlobals newAST (fun (current_global_rename_mapping: method_rename_assumption list) global ->
+    match generate_global_rename_mapping oldMap global with
+      | Some rename_mapping -> current_global_rename_mapping @ [rename_mapping]
```
Comment (sim642) [inline src/incremental/compareCIL.ml:162]: Appending to a list in `foldGlobals` is stupidly inefficient, especially if there are many. Just prepend instead. And if you really need them in the right order, just reverse the list at the end.

File: scripts/test-incremental-multiple.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
+test=$1
```
Comment (sim642) [inline scripts/test-incremental-multiple.sh:1]: What is this script? There's no explanation.

**Comment by sim642** [comment]
Also have a look at the CI failures. Semgrep complains in two places and the incremental test group should be renumbered to something which already doesn't exist (I suppose we might have created such group in the meanwhile ourselves).

**Comment by michael-schwarz** [comment]
Thank you for this PR, I think further enhancing our incrementality in this direction is a very useful thing!

File: src/incremental/compareAST.ml [hunk 4]
```diff
@@ -5,6 +5,41 @@ type global_type = Fun | Decl | Var
 
 and global_identifier = {name: string ; global_t: global_type} [@@deriving ord]
 
+type method_rename_assumption = {original_method_name: string; new_method_name: string; parameter_renames: (string, string) Hashtbl.t}
+type method_rename_assumptions = (string, method_rename_assumption) Hashtbl.t
+
+(*rename_mapping is carried through the stack when comparing the AST. Holds a list of rename assumptions.*)
```
[the remark this answers is not in the document]
Comment (sim642) [reply src/incremental/compareAST.ml:11]: Any global state will make it impossible to parallelize parsing/merging/comparing, so not introducing new global state into these parts would be good, if we hope some day to be able to speed up our preprocessing.

There is another `(typ * typ) list` structure being passed around in many of the comparison functions though, so with this change there would be two. Since most recursion here just passes them around, they could be packaged together into a single record type, which is also easier to extend in the future with other structures.

File: src/analyses/apron/apronAnalysis.apron.ml [hunk 5]
```diff
@@ -146,7 +146,7 @@ struct
     if !GU.global_initialization && e = MyCFG.unknown_exp then
       st (* ignore extern inits because there's no body before assign, so the apron env is empty... *)
     else (
-      if M.tracing then M.traceli "apron" "assign %a = %a\n" d_lval lv d_exp e;
+      if M.tracing then M.traceli "apron" "assign %a = %a\n" RenameMapping.d_lval lv d_exp e;
```
[the remark this answers is not in the document]
Comment (stilscher) [reply src/analyses/apron/apronAnalysis.apron.ml:149]: I was also thinking that, since the constraint system is generated from the new CFG, it is not sufficient to rename a `varinfo` from the old Cil file. But this could actually still work, because of the substitution of the locals, formals, statements and so on, that is done for unchanged functions (in the code snippet I linked above). One only needs to do something similar for the partially changed functions (CFG comparison) also. We should check anyway, why this is not the case yet.

[the remark this answers is not in the document]
Comment (sim642) [reply src/analyses/apron/apronAnalysis.apron.ml:149]: > And I see no point why locations should be handled any different then? Could we not also just change the location of the old data structures to contain the new one?

The fields of `Cil.location` are not mutable right now at least: in `Cil.instr` and `Cil.stmtkind` they are just immutable components of constructor argument tuples. If they were, then it also should be possible, but that requires significant changes to CIL itself.
And the output of locations happens in much fewer places in Goblint code, so the location remapping isn't really noticeable in most code.

File: scripts/test-incremental-multiple.sh (hunk 3 above)
Comment (michael-schwarz) [reply scripts/test-incremental-multiple.sh:1]: That's a very helpful feature indeed! We should maybe start testing all incremental tests in this matter in the CI (with the first patch being the one we have and the second one empty), we currently don't test such multiruns at all automatically.

Comment (sim642) [reply scripts/test-incremental-multiple.sh:1]: There's actually an infinitude of testing scenarios possible for incremental tests:
* applying empty patches,
* reversing the before and after order,
* applying multiple patches in a row,
* pairwise application of those multiple patches,
* doing all the above in server mode instead,
* ...
I'm not sure how beneficial it would be to hack more and more of such interesting incremental testing scenarios into the existing already-convoluted script. They could be feature requests for GobExec (which we should start writing one day) such that we can design nice abstractions from the ground up to allow all these without piling hacks on hacks on hacks.

File: src/analyses/apron/apronAnalysis.apron.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/analyses/apron/apronAnalysis.apron.ml:149]: I have been playing around with simply renaming the varinfos in the old AST as you @sim642 suggested. [Here is my code that performs the renaming](https://github.com/TimOrtel/analyzer/blob/dev-local-var-rename-without-rename-mapping/src/incremental/updateCil.ml#L46). This works in most cases, e.g. in g2html the graph displays the correctly names. However, if you click on the details of each node in g2html, the renamings are not present there, and the old/original names are displayed there. The code responsible for printing the varinfo names in this detail view is [basetype.ml#Variables#show](https://github.com/goblint/analyzer/blob/master/src/cdomains/basetype.ml#L30). 

Probably I am missing something, but to me it looks like that there are at least two copies of the varinfos, because otherwise my renaming should have affected the vname. I am unsure how I can also rename these instances, because I couldn't quite figure out where these are coming from.

[the remark this answers is not in the document]
Comment (michael-schwarz) [reply src/analyses/apron/apronAnalysis.apron.ml:149]: > Probably I am missing something, but to me it looks like that there are at least two copies of the varinfos, because otherwise my renaming should have affected the vname.

Yeah, that looks like it! Just a hunch: what happens if you turn off hashconsing? There should be a switch `ana.opt.hashcons` or something similar...

[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/analyses/apron/apronAnalysis.apron.ml:149]: Running test-incremental with an additional argument --disable ana.opt.hashcons does not change anything in the g2html output.

[the remark this answers is not in the document]
Comment (michael-schwarz) [reply src/analyses/apron/apronAnalysis.apron.ml:149]: ok, was worth a try!

[the remark this answers is not in the document]
Comment (sim642) [reply src/analyses/apron/apronAnalysis.apron.ml:149]: I think the big reason why it doesn't work yet is #357. Right now `incremental.save` will marshal the old `Cil.file` and the solution to separate files. When `incremental.load` unmarshals them, the `varinfo`s in the loaded AST and the loaded solution are no longer shared, so modifying them in the AST doesn't modify them in the solution (including the local values for nodes in g2html).

Marshaling everything into a single file should keep the physical sharing on unmarshaling and allow this to work fully.

[the remark this answers is not in the document]
Comment (jerhard) [reply src/analyses/apron/apronAnalysis.apron.ml:149]: >When `incremental.load` unmarshals them, the `varinfos` in the loaded AST and the loaded solution are no longer shared, so modifying them in the AST doesn't modify them in the solution (including the local values for nodes in g2html).

This should be fixed with #748 being merged into `master`.

File: src/incremental/compareAST.ml (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/compareAST.ml:11]: The carrying through the stack is required for the detection of renamed functions and global variables which I am also working on. As such I am currently working on making the whole AST comparison fully functional without side effects.

File: src/incremental/compareAST.ml [hunk 6]
```diff
@@ -5,6 +5,41 @@ type global_type = Fun | Decl | Var
 
 and global_identifier = {name: string ; global_t: global_type} [@@deriving ord]
 
+type method_rename_assumption = {original_method_name: string; new_method_name: string; parameter_renames: (string, string) Hashtbl.t}
+type method_rename_assumptions = (string, method_rename_assumption) Hashtbl.t
```
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/compareAST.ml:9]: Yes indeed, this is implemented already in a second pull request I will create once this pull request is resolved. Or should I merge theses changes into this pull request already?

File: src/framework/analyses.ml [hunk 7]
```diff
@@ -8,6 +8,8 @@ open GobConfig
 module GU = Goblintutil
 module M  = Messages
 
+let currentFunctionName: string ref = ref ""
```
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/framework/analyses.ml]: You are right, that was bad design. Since it is not needed anymore without GlobalRenameMapping, I have removed it.

File: tests/incremental/04-var-rename/01-nothing.patch [hunk 8]
```diff
@@ -0,0 +1,8 @@
+--- tests/incremental/04-var-rename/01-nothing.c
++++ tests/incremental/04-var-rename/01-nothing.c
+@@ -1,4 +1,5 @@
+ int main() {
+     int x = 0;
++
```
[the remark this answers is not in the document]
Comment (TimOrtel) [reply tests/incremental/04-var-rename/01-nothing.patch]: It's removed.

File: src/incremental/compareAST.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (jerhard) [reply src/incremental/compareAST.ml:9]: Would the second PR introduce new (unrelated?) features? Or would it mostly rework the implementation of this PR?

File: src/incremental/compareAST.ml [hunk 9]
```diff
@@ -5,6 +5,41 @@ type global_type = Fun | Decl | Var
...
+
+let rename_mapping_to_string (rename_mapping: rename_mapping) = 
+  let (local, methods) = rename_mapping in
+  let local_string = string_tuple_to_string (List.of_seq (Hashtbl.to_seq local)) in
+  let methods_string: string = List.of_seq (Hashtbl.to_seq_values methods) |>
+    List.map (fun x -> match x with {original_method_name; new_method_name; parameter_renames} -> 
+      "(methodName: " ^ original_method_name ^ " -> " ^ new_method_name ^ 
+      "; renamed_params=" ^ string_tuple_to_string (List.of_seq (Hashtbl.to_seq parameter_renames)) ^ ")") |>
+    String.concat ", " in
+  "(local=" ^ local_string ^ "; methods=[" ^ methods_string ^ "])"
```
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/compareAST.ml]: Hashtbl does not seem to be supported by deriving show. This is a great tip though, i will try to use it from now on. Thanks.

File: src/incremental/compareAST.ml [hunk 10]
```diff
@@ -5,6 +5,41 @@ type global_type = Fun | Decl | Var
...
+  | Some now -> 
+    (*Printf.printf "Assumption is: %s -> %s\n" original now;*)
+    now = name2
+  | None -> 
+    (*Printf.printf "No assumption when %s, %s, %b\n" name1 name2 (name1 = name2);*)
+    name1 = name2 (*Var names differ, but there is no assumption, so this can't be good*)
+
+let string_tuple_to_string (tuple: (string * string) list) = "[" ^ (tuple |>
+  List.map (fun x -> match x with (first, second) -> "(" ^ first ^ " -> " ^ second ^ ")") |>
+  String.concat ", ") ^ "]"
```
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/compareAST.ml:31]: Oh thank you for the tip, I did not know this syntax.

File: src/incremental/compareCIL.ml (hunk 2 above)
Comment (TimOrtel) [reply src/incremental/compareCIL.ml]: Thank you I implemented that optimization.

File: src/incremental/compareAST.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/compareAST.ml:9]: It introduces detection of function renaming. So yes new features.

**Comment by TimOrtel** [comment]
I have removed RenameMapping and implemented other improvements. Maybe you could give me even more feedback regarding my changes.

File: src/incremental/compareAST.ml (hunk 6 above)
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/compareAST.ml]: I implemented it in this PR now!

**Comment by TimOrtel** [comment]
@sim642 I just see it now you mention it. It must have been an error when I was merging my branches.

**Comment by TimOrtel** [comment]
Now it should actually be removed.

File: src/cdomains/baseDomain.ml [hunk 11]
```diff
@@ -115,7 +115,7 @@ struct
   let printXml f r =
     let e = XmlUtil.escape in
     BatPrintf.fprintf f "<value>\n<map>\n<key>\n%s\n</key>\n%a<key>\n%s\n</key>\n%a<key>\n%s\n</key>\n%a\n<key>\n%s\n</key>\n%a</map>\n</value>\n"
-      (e @@ CPA.name ()) CPA.printXml r.cpa
+      (e @@ (CPA.name ())) CPA.printXml r.cpa
```
Comment (sim642) [inline src/cdomains/baseDomain.ml:118]: There's a number of irrelevant changes like this due to the addition and then removal of `RenameMapping` I guess. If you look at the entire PR's changes on GitHub you can easily find where unnecessary extra parenthesis are added or whitespace is changed in otherwise unchanged code. Would be nice to undo these changes such that this PR doesn't introduce noise to `git blame`.

**Review by sim642 — COMMENTED** [review]
I didn't test whether this works (I'll let @stilscher confirm that), but if it does, then it's certainly nice enough now without `RenameMapping`!

A note to ourselves: given that the commits in this PR have been back and forth changing many places for `RenameMapping` and then undoing it all, once it comes to merging I think it'd be best to squash merge this to avoid those back-and-forths from cluttering the git history.

**Review by stilscher — CHANGES_REQUESTED** [review]
I tested the functionality of the support for detecting renamed local variables on a couple of regression tests (with renamings in the incremental run). The number of changed functions as well as the output and analysis result look good in many cases (f.e. renaming in recursive functions, renaming multiple parameters, pointers, renaming successively, swapping names (is not detected but considered as changed)). 
Things that still need to be fixed, are
- not outputting outdated names of formal parameters
- and the implementation should be changed such that it does not hinder when using the cfg comparison (I described this in more detail in the comment above). 
When renaming static local variables the renaming detection does not help, which is to be expected since these are global variables in CIL. I also noticed, that there are problems when renaming functions that also have a declaration. This is due to the missing grouping of functions and corresponding declarations as described in https://github.com/goblint/analyzer/issues/627. My suggestion is, to fix this not within this PR but as part of the linked issue.

File: src/incremental/updateCil.ml [hunk 12]
```diff
@@ -43,7 +43,7 @@ let update_ids (old_file: file) (ids: max_ids) (new_file: file) (changes: change
   in
   let reset_fun (f: fundec) (old_f: fundec) =
     f.svar.vid <- old_f.svar.vid;
-    List.iter2 (fun l o_l -> l.vid <- o_l.vid) f.slocals old_f.slocals;
+    List.iter2 (fun l o_l -> l.vid <- o_l.vid; o_l.vname <- l.vname) f.slocals old_f.slocals;
     List.iter2 (fun lo o_f -> lo.vid <- o_f.vid) f.sformals old_f.sformals;
```
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/updateCil.ml]: Added!

File: src/incremental/compareCIL.ml [hunk 13]
```diff
@@ -36,18 +37,45 @@ let should_reanalyze (fdec: Cil.fundec) =
...
   let identical, diffOpt =
     if should_reanalyze a then
       false, None
     else
-      let sameDef = unchangedHeader && GobList.equal eq_varinfo a.slocals b.slocals in
+      (* Here the local variables are checked to be equal *)
+      let sizeEqual, local_rename = rename_mapping_aware_compare a.slocals b.slocals headerRenameMapping in
+      let rename_mapping: rename_mapping = (local_rename, global_rename_mapping) in
+
+      let sameDef = unchangedHeader && sizeEqual in
```
[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/compareCIL.ml:73]: I think the second approach is a nice idea. If I still find the time to do it, I will definitely implement it. However, because implementing the second approach comes with a lot of extra work in testing and verifying I implemented the first version for now.

**Comment by TimOrtel** [comment]
Thank you for taking the time to look at my PR @stilscher. I have implemented your feedback.

File: src/incremental/compareCIL.ml (hunk 13 above)
[the remark this answers is not in the document]
Comment (stilscher) [reply src/incremental/compareCIL.ml:72]: I created issue #777 for this.

[the remark this answers is not in the document]
Comment (stilscher) [reply src/incremental/compareCIL.ml:72]: However, your implementation for the first approach does still not work correctly. The comparison of the local variables is still done with the rename-aware comparison even if the cfg comparison is activated. This leads to renamed and unused local variables in a partially changed function not being shown with the updated name in the output. As an example you can take a look at
```
#include<assert.h>

int main () {
  int a = 3;
  int b; // rename to d
  int c;
  c = a + 2; // change to a + 3
  assert(a == 3);
  return 0;
}
```
In the incremental run, the old name `b` will still be used in the output.

[the remark this answers is not in the document]
Comment (stilscher) [reply src/incremental/compareCIL.ml:72]: I actually think, that implementing approach 2 would not be much more work. But I think it is ok, to postpone it and implement in a new PR.

[the remark this answers is not in the document]
Comment (TimOrtel) [reply src/incremental/compareCIL.ml]: Ok, the locals are now checked again for cfg runs.

**Review by stilscher — APPROVED** [review]

**Review by sim642 — APPROVED** [review]
